### PR TITLE
rfc: oneDNN v2 API

### PR DESCRIPTION
**Pull request by emfomenk**
# Description

An RFC on the API changes for the upcoming [oneDNN v2.0](https://github.com/oneapi-src/oneDNN/milestone/7).

[Rendered document](https://github.com/emfomenk/mkl-dnn/tree/emfomenk/rfcs/v2-general-api/rfcs/20200717-onednn-v2#onednn-v20-changes).

The executive summary could be found [here](https://github.com/emfomenk/mkl-dnn/tree/emfomenk/rfcs/v2-general-api/rfcs/20200717-onednn-v2#1-executive-summary-of-the-changes). The API mock up: [link](https://github.com/emfomenk/mkl-dnn/blob/emfomenk/rfcs/v2-general-api/rfcs/20200717-onednn-v2/api-simplified). 

In general, the changes are very modest.
- For oneDNN v1.x users there are very few incompatible changes (listed [here](https://github.com/emfomenk/mkl-dnn/blob/emfomenk/rfcs/v2-general-api/rfcs/20200717-onednn-v2/general-v2-api.md#43-incompatible-changes-v1x-and-v2x)).
- For oneDNN v2.x-beta users, there changes are noticeable but should be manageable (listed [here](https://github.com/emfomenk/mkl-dnn/blob/emfomenk/rfcs/v2-general-api/rfcs/20200717-onednn-v2/general-v2-api.md#44-incompatible-changes-dev-v2-v20-betax-and-v2x)).

Many people contributed to the RFC.
I would like to say special thanks to @echeresh, @rsdubtso, @densamoilov, @sbogusev, and @mgouicem!

### CODE REVIEW

**Comment by emfomenk** [comment]
Cast @kruus, @nSircombe, @kawakami-k, @CaoZhongZ, @pinzhenx

File: rfcs/20200717-onednn-v2/api-simplified/v2_dnnl_threadpool.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#include "dnnl.hpp"
+#include "dnnl_threadpool_iface.hpp"
+
+namespace dnnl {
+namespace threadpool {
+
+// option 1a (preferable, but breaks backwards compatibility)
```
Comment (sbogusev) [inline rfcs/20200717-onednn-v2/api-simplified/v2_dnnl_threadpool.hpp:7]: Just wanted to note, that, as far as I understand, breaking the backwards compatibility here is not an issue. Changes for framework integration may be required for handling threadpool in oneDNN, regardless of potentially dropping stream attributes.

Comment (emfomenk) [reply rfcs/20200717-onednn-v2/api-simplified/v2_dnnl_threadpool.hpp:7]: That's true, however if someone used `stream_attr` (empty one) this code will stop working. So formally we break backwards compatibility even for non-threadpool case. I don't expect anyone does this though... (resolved)

File: rfcs/20200717-onednn-v2/README.md [hunk 2]
```diff
@@ -132,6 +132,13 @@ documents discuss some of the options that we decided not to proceed with:
     more flexible, and in future, if buffers are de-prioritized / dropped,
     there would be minimum changes to the API (essentially 0).
 
+- 20/09/09. After we discovered the issue with namespace clashing, we renamed
+  `dnnl::sycl` (and other) namespaces to `dnnl::sycl_interop`. Along with that
+  we also simplified the names of the interoperability functions. For instance,
+  instead of `dnnl::sycl_interop::memory_get_memory_kind()` the function is now
+  called ``dnnl::sycl_interop::get_memory_kind()` as subject can be guessed by
```
Comment (sbogusev) [inline rfcs/20200717-onednn-v2/README.md:139]:
```suggestion
  called `dnnl::sycl_interop::get_memory_kind()` as subject can be guessed by
```

Comment (emfomenk) [reply rfcs/20200717-onednn-v2/README.md]: Good catch, thanks! Fixed.

**Comment by emfomenk** [comment]
The RFC was approved internally and was on GitHub for quite some time. Assuming reviewed :)

**Pull request merged**